### ISSUE TITLE
Incorrect nullable analysis of `x?.y.Extension() == true` pattern

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#41107]
![image](https://user-images.githubusercontent.com/12466233/72776851-6b6c9e80-3bc8-11ea-8fec-979b29758be2.png)

``` csharp

using System.Diagnostics.CodeAnalysis;

#nullable enable

public class C 
{
    public C? field;
    
    static public void M(C c) 
    {
       if (c.field?.field.IsKind() == true)
       {
          c.field.field.ToString(); // warning CS8602: Dereference of a possibly null reference.
       }
    }
    
}

public static class Extension
{
    public static bool IsKind([NotNullWhen(true)] this C? c)
    {
        throw null!;
    }
}
```

[sharplab repro](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+BYAUAAQAwAEOAjAHQAiAlgIYDmAdhAM4AulYTpAwhACYwBBetQA2ATyaUmAbmzYAxPQCuIkdWAiYBGMI0w5uAMxEATAS4FsAb2wE7RY1wD8BAGaUYI3rKz3Lv+xIkByIUAgBZAAoLMABKfz8bAL8CSlcCSLBSd09eJ2yPL1IASSYAaUp6Xkj4gF5aghYoJRhY2xSCJI6/LJyivt5SABUIAGUmytoa6QIAelmCAHdqKHpJ81GADiQ8ExACchhYVyOdMC0IdOoCAAdmSQ0xAmVVAmPT+nPSdpSAXx+CP9kv4gdgcMYgqYCABRBAsHSSCD0awA8FEYjBYAQCAiAilCpVSIAbQAchAWCSVCIAOoACx0kSaLViAF1GrSpOYXHEAV1uixaVAIItnlSAIQ+PxAoFAA===)

**Comment by jcouv** [comment]
FYI @cston

**Comment by xt0rted** [comment]
I think this is the same issue, but in the scaffolded identity framework code a number of views contain the following pattern which is being flagged:

```csharp
if ((Model.ExternalLogins?.Count ?? 0) == 0)
{
    // ...
}
else
{
    foreach (var provider in Model.ExternalLogins)
    {
        // the ExternalLogins usage above is flagged as possibly null
    }
}
```

If you expand the check to `if (Model.ExternalLogins == null || Model.ExternalLogins.Count == 0)` the warning goes away.

Examples:
- [dotnet/aspnetcore@Login.cshtml#L55-L78](https://github.com/dotnet/aspnetcore/blob/9cab3e2602f540bfd319612d6edcbe3c51876c9d/src/Identity/UI/src/Areas/Identity/Pages/V4/Account/Login.cshtml#L55-L78)
- [dotnet/Scaffolding@Account.Login.cshtml#L55-L78](https://github.com/dotnet/Scaffolding/blob/11006dd65d739c19a80c27029e954ff81e3ac087/src/VS.Web.CG.Mvc/Templates/Identity/Bootstrap4/Pages/Account/Account.Login.cshtml#L55-L78)

**Comment by jaredpar** [comment]
@RikkiGibson another potential definite assignment issue that gets fixed after your feature work.

**Comment by RikkiGibson** [comment]
An upcoming "improved definite assignment" PR fixes this issue.
